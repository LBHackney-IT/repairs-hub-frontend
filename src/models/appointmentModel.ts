export type Appointment = {
  date: string
  description: string
  start: string
  end: string
  reason: string
  note: string
  assignedStart: string
  assignedEnd: string
  startedAt: string
  bookingLifeCycleStatus: string
}
