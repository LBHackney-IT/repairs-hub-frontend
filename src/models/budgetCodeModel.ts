export type BudgetCode = {
  externalCostCode: string
  corporateSubjectiveCode: string
  descriptionOfWorks: string
}
