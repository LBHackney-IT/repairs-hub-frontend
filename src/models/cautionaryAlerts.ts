export type CautionaryAlert = {
  type: string
  comments: string
  startDate: string
  endDate: string
}

export type CautionaryAlertsResponse = {
  reference: string
  alerts: CautionaryAlert[]
}
