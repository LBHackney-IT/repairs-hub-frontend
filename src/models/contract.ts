export default interface Contract {
  contractReference: string
  terminationDate: string
  effectiveDate: string
  contractorReference: string
  isRaisable: boolean
  contractorName: string
  sorCost: number
  sorCount: number
}
