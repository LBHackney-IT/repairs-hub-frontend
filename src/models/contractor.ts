export default interface Contractor {
  contractorReference: string
  contractorName: string
  activeContractCount: number
}