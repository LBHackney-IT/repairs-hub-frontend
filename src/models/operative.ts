export type Operative = {
  jobPercentage: number
  id: number
  payrollNumber: string
  name: string
  trades: string[]
  isOnejobatatime: boolean
}
