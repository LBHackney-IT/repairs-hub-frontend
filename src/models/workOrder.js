import { formatISO, isSameDay } from 'date-fns'
import {
  HIGH_PRIORITY_CODES,
  PRIORITY_CODES_REQUIRING_APPOINTMENTS,
} from '@/utils/helpers/priorities'

import {
  CLOSED_STATUS_DESCRIPTIONS,
  CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES,
} from '@/utils/statusCodes'

export class WorkOrder {
  constructor(workOrderData) {
    Object.assign(this, workOrderData)
  }

  isHigherPriority = () => {
    return HIGH_PRIORITY_CODES.includes(this.priorityCode)
  }

  isAppointmentRequired = () => {
    return PRIORITY_CODES_REQUIRING_APPOINTMENTS.includes(this.priorityCode)
  }

  canBeScheduled = () => {
    return (
      this.statusAllowsScheduling() &&
      this.isAppointmentRequired() &&
      !this.isOutOfHoursGas()
    )
  }

  isOutOfHoursGas = () => {
    const GasBreakdownContractorReference = 'H04'
    const OohTradeCode = 'OO'

    var contractorReference = this.contractorReference

    if (contractorReference != GasBreakdownContractorReference) return false // contractor must be "H04"
    var tradeCode = this.tradeCode
    return tradeCode == OohTradeCode
  }

  completionReason = () => {
    return this.status === 'Work Completed' ? 'Completed' : this.status
  }

  appointmentISODatePassed = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    }

    const currentISODate = formatISO(new Date(), { representation: 'date' })

    const appointmentISODate = formatISO(
      new Date(`${this.appointment.date}T${this.appointment.start}`),
      { representation: 'date' }
    )

    return currentISODate >= appointmentISODate
  }

  appointmentIsToday = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    } else {
      return isSameDay(
        new Date(),
        new Date(`${this.appointment.date}T${this.appointment.start}`)
      )
    }
  }

  statusAllowsScheduling = () => {
    return !CLOSED_STATUS_DESCRIPTIONS.includes(this.status)
  }

  targetTimePassed = () => {
    const currentTime = new Date().getTime()
    const targetTime = new Date(`${this.target}`).getTime()

    return currentTime > targetTime
  }

  hasBeenVisited = () =>
    CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES.includes(this.status)
}
