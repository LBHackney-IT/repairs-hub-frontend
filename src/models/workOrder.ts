import { formatISO, isSameDay } from 'date-fns'
import {
  HIGH_PRIORITY_CODES,
  PRIORITY_CODES_REQUIRING_APPOINTMENTS,
} from '@/utils/helpers/priorities'

import {
  CLOSED_STATUS_DESCRIPTIONS,
  CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES,
} from '@/utils/statusCodes'
import { Operative } from './Operative'
import { Appointment } from './appointmentModel'

export class WorkOrder {
  reference: string
  description: string
  externalAppointmentManagementUrl: string
  startTime: string
  appointment: Appointment
  status: string
  dateRaised: string
  tradeDescription: string
  isFollowOn: boolean
  contractorReference: string
  tradeCode: string
  priorityCode
  target
  budgetCode
  closedDated
  operatives: Operative[]
  owner
  priority: string
  raisedBy: string
  callerName: string
  callerNumber: string

  constructor(workOrderData) {
    Object.assign(this, workOrderData)
  }

  isHigherPriority = () => {
    return HIGH_PRIORITY_CODES.includes(this.priorityCode)
  }

  isAppointmentRequired = () => {
    return PRIORITY_CODES_REQUIRING_APPOINTMENTS.includes(this.priorityCode)
  }

  canBeScheduled = () => {
    return (
      this.statusAllowsScheduling() &&
      this.isAppointmentRequired() &&
      !this.isOutOfHoursGas()
    )
  }

  isOutOfHoursGas = () => {
    const gasBreakdownContractorReference = 'H04'
    const oohTradeCode = 'OO'

    const contractorReference = this.contractorReference

    if (contractorReference != gasBreakdownContractorReference) return false // contractor must be "H04"
    const tradeCode = this.tradeCode
    return tradeCode == oohTradeCode
  }

  completionReason = () => {
    return this.status === 'Work Completed' ? 'Completed' : this.status
  }

  appointmentISODatePassed = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    }

    const currentISODate = formatISO(new Date(), { representation: 'date' })

    const appointmentISODate = formatISO(
      new Date(`${this.appointment.date}T${this.appointment.start}`),
      { representation: 'date' }
    )

    return currentISODate >= appointmentISODate
  }

  appointmentIsToday = () => {
    if (
      !this.appointment ||
      !this.appointment.date ||
      !this.appointment.start
    ) {
      return false
    } else {
      return isSameDay(
        new Date(),
        new Date(`${this.appointment.date}T${this.appointment.start}`)
      )
    }
  }

  statusAllowsScheduling = () => {
    return !CLOSED_STATUS_DESCRIPTIONS.includes(this.status)
  }

  targetTimePassed = () => {
    const currentTime = new Date().getTime()
    const targetTime = new Date(`${this.target}`).getTime()

    return currentTime > targetTime
  }

  hasBeenVisited = () =>
    CLOSED_STATUS_DESCRIPTIONS_FOR_OPERATIVES.includes(this.status)
}
