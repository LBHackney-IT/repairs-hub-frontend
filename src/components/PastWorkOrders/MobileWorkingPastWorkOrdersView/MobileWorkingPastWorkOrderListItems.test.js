import React from 'react'
import { WorkOrder } from '../../../models/workOrder'
import { operative } from 'factories/operative'
import { render, screen } from '@testing-library/react'
import { MobileWorkingPastWorkOrderListItems } from './MobileWorkingPastWorkOrderListItems'

const mockWorkOrders = [
  {
    appointment: {
      date: '2024-11-20',
      description: 'Externally Managed Appointment',
      start: '08:00',
      end: '16:15',
      reason: 'FIRST',
      note: null,
      assignedStart: '11:05',
      assignedEnd: '13:01',
      startedAt: '11:05',
    },
    reference: 1,
    dateRaised: '2024-11-18T11:48:21.276019Z',
    lastUpdated: null,
    priority: '[E] EMERGENCY',
    priorityCode: null,
    property: '1 Hackney',
    propertyPostCode: 'N16',
    owner: 'HH General Building Repair',
    description: 'Work order 1',
    propertyReference: '00030674',
    tradeCode: 'PL',
    tradeDescription: 'Plumbing - PL',
    status: 'Completed',
    drsSyncStatus: null,
    timeOrderSentToDrs: null,
  },
  {
    appointment: {
      date: '2024-11-20',
      description: 'Externally Managed Appointment',
      start: '08:00',
      end: '16:15',
      reason: 'FIRST',
      note: null,
      assignedStart: '11:05',
      assignedEnd: '13:01',
      startedAt: '11:05',
    },
    reference: 2,
    dateRaised: '2024-11-18T11:48:21.276019Z',
    lastUpdated: null,
    priority: '[E] EMERGENCY',
    priorityCode: null,
    property: '1 Hackney',
    propertyPostCode: 'N16',
    owner: 'HH General Building Repair',
    description: 'Work order 2',
    propertyReference: '00030674',
    tradeCode: 'PL',
    tradeDescription: 'Plumbing - PL',
    status: 'Completed',
    drsSyncStatus: null,
    timeOrderSentToDrs: null,
  },
]

describe('MobileWorkingPastWorkOrderListItems Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should match snapshot when there are work orders', () => {
    const { asFragment } = render(
      <MobileWorkingPastWorkOrderListItems
        currentUser={operative}
        workOrders={mockWorkOrders.map((wo) => new WorkOrder(wo))}
      />
    )
    expect(screen.getByText('Work order 1'))
    expect(screen.getByText('Work order 2'))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should be empty when there are no orders', () => {
    const { asFragment } = render(
      <MobileWorkingPastWorkOrderListItems
        currentUser={operative}
        workOrders={[]}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
