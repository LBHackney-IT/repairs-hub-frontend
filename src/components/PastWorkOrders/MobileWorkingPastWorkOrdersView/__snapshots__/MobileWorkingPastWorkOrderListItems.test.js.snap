// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileWorkingPastWorkOrderListItems Component should be empty when there are no orders 1`] = `<DocumentFragment />`;

exports[`MobileWorkingPastWorkOrderListItems Component should match snapshot when there are work orders 1`] = `
<DocumentFragment>
  <li
    class="govuk-!-margin-top-3 operative-work-order-list-item operative-work-order-list-item--inactive"
    data-id="0"
    style="cursor: pointer;"
  >
    <div
      class="appointment-details"
    >
      <div>
        <h3
          class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-display-inline"
        >
          08:00 – 16:15
        </h3>
        <span
          class="govuk-tag lbh-tag lbh-tag--status-completed work-order-status govuk-!-margin-top-0 govuk-!-margin-left-2"
        >
          Completed
        </span>
      </div>
      <p
        class="lbh-body lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-margin-bottom-2 capitalize"
      >
        [e] emergency
      </p>
      <h3
        class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-display-inline"
      >
        WO 1
      </h3>
      <p
        class="lbh-body govuk-!-margin-0"
      >
        1 Hackney
      </p>
      <p
        class="lbh-body govuk-!-margin-0 govuk-!-margin-bottom-8"
      >
        N16
      </p>
      <p
        class="lbh-body govuk-!-margin-0 truncate-description truncate-line-3"
      >
        Work order 1
      </p>
    </div>
    <div
      class="govuk-!-margin-0"
    >
      <span
        class="arrow right"
      />
    </div>
  </li>
  <li
    class="govuk-!-margin-top-3 operative-work-order-list-item operative-work-order-list-item--inactive"
    data-id="1"
    style="cursor: pointer;"
  >
    <div
      class="appointment-details"
    >
      <div>
        <h3
          class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-display-inline"
        >
          08:00 – 16:15
        </h3>
        <span
          class="govuk-tag lbh-tag lbh-tag--status-completed work-order-status govuk-!-margin-top-0 govuk-!-margin-left-2"
        >
          Completed
        </span>
      </div>
      <p
        class="lbh-body lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-margin-bottom-2 capitalize"
      >
        [e] emergency
      </p>
      <h3
        class="lbh-heading-h3 lbh-!-font-weight-bold govuk-!-margin-0 govuk-!-display-inline"
      >
        WO 2
      </h3>
      <p
        class="lbh-body govuk-!-margin-0"
      >
        1 Hackney
      </p>
      <p
        class="lbh-body govuk-!-margin-0 govuk-!-margin-bottom-8"
      >
        N16
      </p>
      <p
        class="lbh-body govuk-!-margin-0 truncate-description truncate-line-3"
      >
        Work order 2
      </p>
    </div>
    <div
      class="govuk-!-margin-0"
    >
      <span
        class="arrow right"
      />
    </div>
  </li>
</DocumentFragment>
`;
